### PR TITLE
Build instruction profiler runtime as part of compiler-rt

### PR DESCRIPTION
**Pull request by marco-c**
r? @alexcrichton 

This is #38608 with some fixes.

Still missing:
- [x] testing with profiler enabled on some builders (on which ones? Should I add the option to some of the already existing configurations, or create a new configuration?);
- [x] enabling distribution (on which builders?);
- [x] documentation.

### CODE REVIEW

**Comment by eddyb** [comment]
cc @whitequark

**Comment by kennytm** [comment]
I'm just wondering, aside from the name, is there any difference between `#![sanitizer_runtime]` and `#![profiler_runtime]`, that leads to a distinct set of features?

**Comment by alexcrichton** [comment]
Looking great, thanks for picking this up @marco-c!


> testing with profiler enabled on some builders (on which ones? Should I add the option to some of the already existing configurations, or create a new configuration?);

I think this is supported on all platforms, right? If that's the case you can add `--enable-sanitizers` [here](https://github.com/rust-lang/rust/blob/d015610db7b7d30f3c92d3272ce166397a81349e/src/ci/run.sh#L30-L34). If not you can selectively enable it in particular builders, for example for `x86_64-gnu` you'd enable it [here](https://github.com/rust-lang/rust/blob/d015610db7b7d30f3c92d3272ce166397a81349e/src/ci/docker/x86_64-gnu/Dockerfile#L25) next to `--enable-sanitizers`. For OSX you'd do it [here](https://github.com/rust-lang/rust/blob/d015610db7b7d30f3c92d3272ce166397a81349e/.travis.yml#L57) and on Windows  you'd do it [here](https://github.com/rust-lang/rust/blob/d015610db7b7d30f3c92d3272ce166397a81349e/appveyor.yml#L10).

> enabling distribution (on which builders?)

This sort of depends on the point above. If we want to enable this on all platforms then `src/ci/run.sh` will do the trick. If not let's work out which builders need to distribute the profiler and we can figure that out.

> documentation.

FWIW I'm not thinking anything fancy here, just a "hello world" documentation thing inside of src/doc/unstable-book/src/compiler-flags/profile.md.

**Comment by marco-c** [comment]
I've added some basic documentation, without going into detail into how to parse the gcno/gcda files, as I thought it was out of scope (if you think it isn't, I can describe in more detail how to generate a report using lcov and genhtml).

**Comment by marco-c** [comment]
> I'm just wondering, aside from the name, is there any difference between #![sanitizer_runtime] and #![profiler_runtime], that leads to a distinct set of features?

The profiler instruments the binary to collect coverage data, the sanitizers (there are many different sanitizers) are mostly used to detect memory/threading errors. The sanitizers can also be used to collect coverage data (https://clang.llvm.org/docs/SanitizerCoverage.html), but the output isn't compatible with gcov that is the most widely used tool.

**Comment by alexcrichton** [comment]
Ah it looks like maybe this won't work with a system and/or older LLVM? Travis failed with:

```
[00:46:39] LLVM ERROR: IO failure on output stream.
```

File: src/doc/unstable-book/src/compiler-flags/profile.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# `profile`
+
+The tracking issue for this feature is: None
+
+------------------------
+
+This feature allows the generation of code coverage reports.
+
+Set the compiler flags `-Ccodegen-units=1 -Clink-dead-code -Cpasses=insert-gcov-profiling -Zno-landing-pads` to enable gcov profiling.
```
Comment (alexcrichton) [inline src/doc/unstable-book/src/compiler-flags/profile.md]: Shouldn't `-Z profile` be mentioned here? If all these other flags are required, could `-Z profile` imply them?

**Comment by kennytm** [comment]
@marco-c Yes I understand the difference between sancov and gcov. But what I mean is, `#![sanitizer_runtime]` and `#![profiler_runtime]` only affect linker behavior ("link to this crate if `-Zsanitizer=X`/`-Zprofile` is used"). The reasons of split I can see are:

* The profiler runtime is always a staticlib, while the sanitizer runtime are sometimes dylib
* You can use `-Zprofile` and `-Zsanitizer=X` simultaneously

**Comment by kennytm** [comment]
@alexcrichton I think it is because [`/checkout` is read-only](https://github.com/rust-lang/rust/blob/ae7920153331f26f2bc144f62082e95e74c2ba31/src/ci/docker/run.sh#L72) on Docker. Those `*.gcno` and `*.gcda` are placed "next to the existing source", so LLVM failed by writing to a read-only drive.

**Comment by marco-c** [comment]
> @alexcrichton I think it is because /checkout is read-only on Docker. Those *.gcno and *.gcda are placed "next to the existing source" (/checkout/src/test/run-make/profile/), so LLVM failed by writing to a read-only drive.

They should be placed in the output directory.
If you do:
```
mkdir target
rustc -g -Z profile main.rs --out-dir target
```
Both the binary and the gcno file will appear in the `target` directory.

The error does definitely seem related to failure in writing files though.

File: src/doc/unstable-book/src/compiler-flags/profile.md (hunk 1 above)
Comment (marco-c) [inline src/doc/unstable-book/src/compiler-flags/profile.md]: Right, I think they aren't strictly required. I've replaced them with `-Zprofile`.

**Comment by marco-c** [comment]
Oh, wait a second, the builder has LLVM 3.7, but the three-element variant of the llvm.gcov metadata is only supported from LLVM 3.9.

@alexcrichton should I revert back to the old two-element variant that is also supported on LLVM < 3.9?
Or should a newer LLVM be used on the builders?

**Comment by whitequark** [comment]
@marco-c see https://github.com/rust-lang/rust/pull/38608#discussion_r113059092 re: LLVM version...

**Comment by alexcrichton** [comment]
Ok in that case we won't want to blanket enable this across all builders (as this is known to fail on at least one).

Let's enable three testing builders for OSX/Windows/Linux as [mentioned above](https://github.com/rust-lang/rust/pull/42433#issuecomment-306269739) and then let's similarly modify the builders that produce releases for these platforms. You can do that by modifying `dist-i686-gnu` and `dist-x86_64-gnu` docker images, [32-bit osx](https://github.com/rust-lang/rust/blob/e1293ecbeaa6996d79282a79bd6f3340b487dab8/.travis.yml#L90), [64-bit osx](https://github.com/rust-lang/rust/blob/e1293ecbeaa6996d79282a79bd6f3340b487dab8/.travis.yml#L104), and [windows deployments](https://github.com/rust-lang/rust/blob/e1293ecbeaa6996d79282a79bd6f3340b487dab8/appveyor.yml#L47-L72)

**Comment by alexcrichton** [comment]
Can you change `dist-x86_64-gnu` and `dist-i686-gnu` in the `src/ci/docker` folder as well? (those are the Linux releases)

**Comment by marco-c** [comment]
> Can you change dist-x86_64-gnu and dist-i686-gnu in the src/ci/docker folder as well? (those are the Linux releases)

I changed `dist-x86_64-linux` and `dist-i686-linux`, as `dist-x86_64-gnu` and `dist-i686-gnu` don't exist :)

**Comment by alexcrichton** [comment]
Oops sorry I missed that! Yeah as-is is fine

**Comment by alexcrichton** [comment]
Ok great! Let's see what @bors has to say...

@bors: r+

**Comment by kennytm** [comment]
x86_64-pc-windows-gnu:

```
cargo:warning=../compiler-rt/lib/profile\InstrProfilingUtil.c:26:25: fatal error: sys/utsname.h: No such file or directory
cargo:warning= #include <sys/utsname.h>
cargo:warning=                         ^
```

**Comment by marco-c** [comment]
The build on Windows is failing because sys/utsname.h is being included (https://github.com/llvm-mirror/compiler-rt/blob/master/lib/profile/InstrProfilingUtil.c#L25), because COMPILER_RT_HAS_UNAME is defined (https://github.com/llvm-mirror/compiler-rt/blob/835a29632377870f0d7863f0101e1be441327476/lib/profile/InstrProfilingPort.h#L48).

I suppose we could avoid enabling the profiler on the GNU Windows targets.

**Comment by aidanhs** [comment]
```
[01:19:54]   = note: libprofiler_builtins-72c8829264981064.rlib(GCDAProfiling.o) : error LNK2019: unresolved external symbol fdopen referenced in function llvm_gcda_start_file
[01:19:54]           libprofiler_builtins-72c8829264981064.rlib(GCDAProfiling.o) : error LNK2019: unresolved external symbol strdup referenced in function mangle_filename
[01:19:54]           libprofiler_builtins-72c8829264981064.rlib(GCDAProfiling.o) : error LNK2019: unresolved external symbol open referenced in function llvm_gcda_start_file
[01:19:54]           libprofiler_builtins-72c8829264981064.rlib(GCDAProfiling.o) : error LNK2019: unresolved external symbol mmap referenced in function map_file
[01:19:54]           libprofiler_builtins-72c8829264981064.rlib(GCDAProfiling.o) : error LNK2019: unresolved external symbol munmap referenced in function unmap_file
[01:19:54]           libprofiler_builtins-72c8829264981064.rlib(GCDAProfiling.o) : error LNK2019: unresolved external symbol msync referenced in function unmap_file
[01:19:54]           libprofiler_builtins-72c8829264981064.rlib(GCDAProfiling.o) : error LNK2019: unresolved external symbol flock referenced in function llvm_gcda_end_file
[01:19:54]           C:/projects/rust/build/x86_64-pc-windows-msvc/test/run-make/profile.stage2-x86_64-pc-windows-msvc\test.exe : fatal error LNK1120: 7 unresolved externals
```
Seems like a legit failure

**Comment by whitequark** [comment]
`WindowsMMap.c` is missing from the set of build targets.

File: src/doc/unstable-book/src/compiler-flags/profile.md [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# `profile`
+
+The tracking issue for this feature is: None
```
Comment (alexcrichton) [inline src/doc/unstable-book/src/compiler-flags/profile.md]: Ah actually can you fill this in with https://github.com/rust-lang/rust/issues/42524?

File: src/doc/unstable-book/src/language-features/profiler-runtime.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# `profiler_runtime`
+
+The tracking issue for this feature is: None.
```
Comment (alexcrichton) [inline src/doc/unstable-book/src/language-features/profiler-runtime.md]: (this'd also get linked to https://github.com/rust-lang/rust/issues/42524)

**Comment by marco-c** [comment]
I added WindowsMMap.c to the list of files to compile and defined `strdup`, `open` and `fdopen` as `_strdup`, `_open` and `_fdopen`.

**Comment by kennytm** [comment]
distcheck failed.

```
[01:23:38] Distcheck rust-src
[01:23:38] error: failed to load source for a dependency on `profiler_builtins`
[01:23:38] 
[01:23:38] Caused by:
[01:23:38]   Unable to update file:///checkout/obj/build/tmp/distcheck-src/rust-src/lib/rustlib/src/rust/src/libprofiler_builtins
[01:23:38] 
[01:23:38] Caused by:
[01:23:38]   failed to read `/checkout/obj/build/tmp/distcheck-src/rust-src/lib/rustlib/src/rust/src/libprofiler_builtins/Cargo.toml`
[01:23:38] 
[01:23:38] Caused by:
[01:23:38]   No such file or directory (os error 2)
```

**Comment by est31** [comment]
```
The job exceeded the maximum time limit for jobs, and has been terminated.
```

Maybe a retry will fix it.

@bors retry

**Comment by est31** [comment]
Spurious musl segfault (#38618)

@bors retry

**Comment by sylvestre** [comment]
Clap clap! 👍

**Comment by ofek** [comment]
So now that this is merged, how can one test with coverage?

**Comment by kennytm** [comment]
@ofek Same as https://jbp.io/2017/07/19/measuring-test-coverage-of-rust-programs, but use `-Zprofile` instead of `-Cpasses=insert-gcov-profiling -L/usr/lib/llvm-3.8/lib/clang/3.8.1/lib/linux/ -lclang_rt.profile-x86_64`.

**Comment by ofek** [comment]
@kennytm Brilliant, thanks!

**Pull request merged**